File: step_05_strings/5.2_medium/IB048_compare_version_numbers/solution.py
"""Problem: Compare Version Numbers - Compare two version strings like "1.2.3" vs "1.3" """

# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(max(n,m))  |  Space: O(n+m)
# Split both strings by ".", pad shorter list with "0"s,
# convert to ints and compare element by element.
# ============================================================
def brute_force(version1: str, version2: str) -> int:
    v1 = list(map(int, version1.split(".")))
    v2 = list(map(int, version2.split(".")))
    # Pad with zeros
    length = max(len(v1), len(v2))
    v1 += [0] * (length - len(v1))
    v2 += [0] * (length - len(v2))
    for a, b in zip(v1, v2):
        if a < b: return -1
        if a > b: return 1
    return 0


# ============================================================
# APPROACH 2: OPTIMAL - Two-pointer, parse segments on the fly
# Time: O(max(n,m))  |  Space: O(1) extra
# Use two string indices; parse each integer segment without splitting
# ============================================================
def optimal(version1: str, version2: str) -> int:
    i, j = 0, 0
    n1, n2 = len(version1), len(version2)
    while i < n1 or j < n2:
        num1 = num2 = 0
        while i < n1 and version1[i] != '.':
            num1 = num1 * 10 + int(version1[i]); i += 1
        while j < n2 and version2[j] != '.':
            num2 = num2 * 10 + int(version2[j]); j += 1
        if num1 < num2: return -1
        if num1 > num2: return 1
        i += 1; j += 1
    return 0


# ============================================================
# APPROACH 3: BEST - zip_longest for clean Pythonic solution
# Time: O(max(n,m))  |  Space: O(n+m)
# Use itertools.zip_longest to handle different lengths cleanly
# ============================================================
def best(version1: str, version2: str) -> int:
    from itertools import zip_longest
    parts1 = version1.split(".")
    parts2 = version2.split(".")
    for a, b in zip_longest(parts1, parts2, fillvalue="0"):
        va, vb = int(a), int(b)
        if va < vb: return -1
        if va > vb: return 1
    return 0


if __name__ == "__main__":
    print("=== Compare Version Numbers ===")
    tests = [
        ("1.0",   "1.0.0", 0),
        ("0.1",   "1.1",  -1),
        ("1.0.1", "1",     1),
        ("1.01",  "1.001", 0),
    ]
    for v1, v2, expected in tests:
        b  = brute_force(v1, v2)
        o  = optimal(v1, v2)
        be = best(v1, v2)
        print(f'"{v1}" vs "{v2}" => Brute={b}, Optimal={o}, Best={be} (exp={expected})')
        assert b == o == be == expected, f"Mismatch: {b} {o} {be}"
    print("All tests passed.")

File: step_05_strings/5.2_medium/IB048_compare_version_numbers/Solution.java
import java.util.*;

// ============================================================
// APPROACH 1: BRUTE FORCE
// Time: O(max(n,m))  |  Space: O(n+m)
// Split both strings by ".", pad shorter one with "0"s,
// compare each corresponding integer token left to right.
// ============================================================
class BruteForce {
    public static int solve(String version1, String version2) {
        String[] v1 = version1.split("\\.");
        String[] v2 = version2.split("\\.");
        int n = Math.max(v1.length, v2.length);
        for (int i = 0; i < n; i++) {
            int n1 = (i < v1.length) ? Integer.parseInt(v1[i]) : 0;
            int n2 = (i < v2.length) ? Integer.parseInt(v2[i]) : 0;
            if (n1 < n2) return -1;
            if (n1 > n2) return 1;
        }
        return 0;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL - Two-pointer without extra array
// Time: O(max(n,m))  |  Space: O(1) extra (ignoring output)
// Use two pointers on each string, parse each segment on the fly.
// ============================================================
class Optimal {
    public static int solve(String version1, String version2) {
        int i = 0, j = 0;
        int n1 = version1.length(), n2 = version2.length();
        while (i < n1 || j < n2) {
            int num1 = 0, num2 = 0;
            while (i < n1 && version1.charAt(i) != '.') num1 = num1 * 10 + (version1.charAt(i++) - '0');
            while (j < n2 && version2.charAt(j) != '.') num2 = num2 * 10 + (version2.charAt(j++) - '0');
            if (num1 < num2) return -1;
            if (num1 > num2) return 1;
            i++; j++;
        }
        return 0;
    }
}

// ============================================================
// APPROACH 3: BEST - Split + Integer.parseInt with leading zero handling
// Time: O(max(n,m))  |  Space: O(n+m)
// Cleanest Java approach; Integer.parseInt handles leading zeros ("001" -> 1)
// ============================================================
class Best {
    public static int solve(String version1, String version2) {
        String[] a = version1.split("\\.", -1);
        String[] b = version2.split("\\.", -1);
        int len = Math.max(a.length, b.length);
        for (int i = 0; i < len; i++) {
            int va = (i < a.length) ? Integer.parseInt(a[i]) : 0;
            int vb = (i < b.length) ? Integer.parseInt(b[i]) : 0;
            if (va != vb) return va < vb ? -1 : 1;
        }
        return 0;
    }
}

public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Compare Version Numbers ===");

        String[][] tests = {
            {"1.0", "1.0.0"},  // expected 0
            {"0.1", "1.1"},    // expected -1
            {"1.0.1", "1"},    // expected 1
            {"1.01", "1.001"}, // expected 0 (leading zeros)
        };
        int[] expected = {0, -1, 1, 0};

        for (int t = 0; t < tests.length; t++) {
            String v1 = tests[t][0], v2 = tests[t][1];
            int bf = BruteForce.solve(v1, v2);
            int op = Optimal.solve(v1, v2);
            int be = Best.solve(v1, v2);
            System.out.printf("\"%s\" vs \"%s\" => Brute=%d, Optimal=%d, Best=%d (exp=%d)%n",
                v1, v2, bf, op, be, expected[t]);
        }
    }
}
